### PR TITLE
Better ejection

### PR DESCRIPTION
**Pull request by OliverW10**
ejects opponent balls out of the shooter at a low speed if it can. This is to prevent the large disruption caused by picking up an opponent ball and having to eject it out the intake, making us unable to intake for a second or two.

### CODE REVIEW

File: controllers/shooter.py [hunk 1]
```diff
@@ -57,6 +58,7 @@ class ShooterController(StateMachine):
     AUTO_MAX_ACCEL = 0.2  # G
     AUTO_ALLOWABLE_TURRET_ERROR = 0.3  # m, ring is 1.22m diameter
 
+    _wants_to_clear = False
     _wants_to_fire = will_reset_to(False)
```
Comment (auscompgeek) [inline controllers/shooter.py:62]: What happens if both of these are true?

I think that is probably undesirable, and we should make that state impossible. For example we could have an enum of possible commands for the shooter.

**Comment by james-ward** [comment]
This is a major departure from the strategy that we all agreed upon. As such, perhaps we should consider what it actually looks like in detail.

For example, if we pick up an opponent ball and spit it out the turret like this, we might infer that we are chewing through a big patch of balls that we can't easily separate. Maybe the intake shouldn't retract if the colour is wrong and it is being ejected through the shooter so that we can keep going until we get our coloured ball?

If we are picking up a ball to catflap, that's more deliberate, so the intake probably should come back up.

Thoughts?

**Comment by OliverW10** [comment]
Yes, this and the ability to toggle auto-retract I added in #139 go hand in hand. I think the auto-retract is only useful situationally (e.g. picking up against walls where you dont have a clear view or when under heavy defense) but otherwise I think raising the intake should be up to driver input.
With auto retract off it will still stop the intake motor running when we have a ball in the tunnel and if it needs to eject it out the front it will raise intake.

**Comment by james-ward** [comment]
Could we come up with a compromise here? I really don't like leaving the intake down all the time because every foul for breaching another robot's frame perimeter costs us the same as two cargo, and if we lose the intake, we are knackered.

How about this (which I think fits nicely with this new clearing approach):
The auto retract only happens when we have two cargo on board. The opposition cargo would be ejected and the intake stays down, but once we have two we get the intake out the way. If the driver is in the corner looking to fill up, the intake will stay down until they have a full load, or they pull it up after one.

**Comment by james-ward** [comment]
Another option would be to not have the override be a toggle, but make a button press be auto retract on each correct ball (stay down for rejections) and holding the button down be "keep it down until full" mode.

File: controllers/indexer.py [hunk 2]
```diff
@@ -60,35 +62,49 @@ def intaking(self) -> None:
...
+            if self.indexer.has_cargo_in_chimney():
+                self.next_state("intake_clearing")
+            else:
+                self.shooter_control.clear()
+
+        if not (
+            self.shooter_control.current_state == "prepare_to_clear"
+            or self.shooter_control.current_state == "clearing"
+        ):
+            self.next_state("stopped")
```
Comment (james-ward) [inline controllers/indexer.py:101]: Will this cause it to never clear through the shooter?
 * Initial call happens
 * Shooter controller set to clear -> this just sets a private flag in the controller
 * Lines 97-100 check the state of the shooter controller, but it hasn't changed yet, so the indexer moves to the `stopped` state

**Review by james-ward — CHANGES_REQUESTED** [review]

**Comment by james-ward** [comment]
Could this be approached differently? From the indexer point of view, clearing through the shooter is the same as firing. So why not keep the states essentially the same, just have a way to override the shooter setpoint speed to the low dribble-it-out speed.

**Comment by james-ward** [comment]
And as I look at all this stuff and think about how the indexer state machine works and how it transfers cargo, I'm now convinced that having a driver override for the intake staying down and turning is a recipe for disaster. The potential to end up with two cargo in the same segment of the indexer is too much. At the very least the intake wheels need to be stopped until the cargo has been moved out of the tunnel and into the chimney.

**Comment by auscompgeek** [comment]
> The auto retract only happens when we have two cargo on board.

I think we might want to have have two auto-retract modes then:

- Auto retract once we have two cargo, like you suggest. This can be used when we explicitly want to pick up two cargo.
- Auto retract whenever we pick up cargo (like the robot current does). Particularly useful when we anticipate heavy defence, or we don't have a clear view of the robot.

Obviously, these two modes would be equivalent once we cat-flap a cargo.

> The opposition cargo would be ejected and the intake stays down

In our testing on Saturday, we found that the intake needs to be raised before we can spit a ball back out the intake.

> Another option would be to not have the override be a toggle, but make a button press be auto retract on each correct ball (stay down for rejections) and holding the button down be "keep it down until full" mode.

I definitely think holding the button down to activate is a good idea. Makes it almost impossible for the person to forget it's active.

**Comment by OliverW10** [comment]
This is the logic i've implemented (if i haven't made a mistake)
 - Thumb button toggles intake, if its down it will retract with any ball
 - Holding button 11 is the 'intake untill full' button, it will keep intake down until it has two balls
 - If there is a ball in the tunnel, intake motor stops, regardless of mode
 - If there is no ball in the chimney and we have an opponent ball it ejects it out the turret
 - If we eject out of the intake it will raise it first

**Comment by james-ward** [comment]
@OliverW10 take a look at #169 for what I think is a nicer way to do the turret rejection bit. Doesn't require changing the indexer state machine states at all.

**Comment by OliverW10** [comment]
Ive also cleaned up the logic in the reading state

**Comment by james-ward** [comment]
This looks good, but we still need to stop the intake rollers when the auto retract is off and we are processing a ball. We cannot allow two balls into the indexer tunnel at the same time.
Same applies to rejecting a ball through the intake with it down - will need to be back driven.

**Comment by OliverW10** [comment]
[the intake rollers stop if there is a ball in the tunnel](https://github.com/thedropbears/pyrapidreact/pull/157/files#diff-a06acd8fb6a9bd3deaa5ec37eba793f8d165afeceb31ab4444018e4cb01b86f5R195) and the [clearing through intake raises intake](https://github.com/thedropbears/pyrapidreact/pull/157/files#diff-1286f76b84283522f93636b5c521e8381f64c5149449c8259df139708160d10eR89), should I make it backdrive intake instead?

**Comment by james-ward** [comment]
I think so. Isn't the idea here to try to save time, which the intake raising and lowering is the biggest time sink. And if the driver is holding the intake down, shouldn't we keep it down even for a rejection?
If we want to raise it regardless when we reject that's fine by me. I just want to be sure that we don't end up in a race condition between the driver inputs and the state machine.

**Comment by james-ward** [comment]
And the reason I didn't see the motor being stopped is that we now have intake operational logic in robot.py
This has become complex enough that I don't think that's appropriate any more.

**Comment by OliverW10** [comment]
I've made it so intake backdrives and also changed where intake is controlled

**Review by james-ward — APPROVED** [review]
lgtm
@auscompgeek

**Pull request merged**